### ISSUE TITLE
Window is not placed in correct position after Maximize/Restore

### ISSUE THREAD

**Issue by gh658804** [RamonUnch/AltSnap#124]
Drag a window by pressing Alt (or whatever key you have assigned) and left-clicking on the window. At some point, without releasing the left button, if you press the right button, the window gets maximized. Cool. Now press the right button again. The window is restored but not placed at the location it was before it was maximized, but it is centered in the screen.

This is a one-step process, the left button is never released. IMO in that case, the position in which the restored window appears should not be "centered" but whatever it was before maximizing it.

**Comment by RamonUnch** [comment]
This is intentional, the window is not restored centered but at the position where the cursor is like when you drag it out of the full screen state. If you want window position to be preserved you should use Alt+double-click or the Maximize action. The point of this option (that you can disable in the advanced tab of the config) is be able to maximize restore windows across monitors while dragging the mouse, so I do not intend to modify it as it would break many workflows. and any way, there is no way I would allow to have a window in a move state without having the cursor inside it, hence I cannot change the behavior.

**Comment by gh658804** [comment]
OK, i think i understand your points, however let me explain why it felt wrong in the first place and led me to report it as a problem.

When i rightclicked to maximize and then i rightclicked again to restore, i had not moved the mouse at all, so i saw the window positioned in a different place not only relative to the old position but also to the position of the mouse (hopefully, its clear what i mean).

As an example, the mouse was lets say at 75% of the window, both horizontally and vertically. After i performed the maximize/restore action, the mouse was now at 50% of the window, both horizontally and vertically.

Having said all that, may i propose a new setting (if you feel it would not be nice to have it as default behavior) that when checked, will position the window in such a place so that the relative position of the mouse inside the window remains the same as before the maximize/restore action.

**Comment by RamonUnch** [comment]
This would be annoying to implement as I would need two ways to calculate the offset relative to the current position.
For now when you restore a maximized window while in the moving state, the position of the windows is calculated so that the cursor position remains at the same point inside the window (in proportion) from its maximized state and not its restored state, changing this would be a mess for now, so I will see later

**Comment by gh658804** [comment]
I see. I feared that it wouldn't be so easy...

Anyway, if possible in the future for the last part of your post to be done, it would be great.

Thanks for your time.

**Comment by RamonUnch** [comment]
After more refactoring it will be easier to have more options without taking the risk to add more bugs. Next release will focus on bug fixes and will add very little features. I still have a crash on Win11, I have no clue on how to fix (I do not have Win 11 so this is a pure guessing job). Features do come eventually, do not worry, One year ago I did not think I would implement any kind of FancyZones integration, but now it is there. This issue will remain open until I fix it.

**Comment by RamonUnch** [comment]
@gh658804 
Here is a version where the window is restored at the same point if no drag occurred when using left+right click:
ie: If you use the click combo (move+resize buttons) WITHOUT moving between the two clicks, then the window will be restored at its previous position. Otherwise it will behave as before, so dragging out a maximized window behaves as before and sets the window under the cursor.

[AltSnap1.52bin.zip](https://github.com/RamonUnch/AltSnap/files/8148496/AltSnap1.52bin.zip)
[AltSnap1.52bin_x64.zip](https://github.com/RamonUnch/AltSnap/files/8148497/AltSnap1.52bin_x64.zip)

Tell me if it fixes it for you

**Comment by gh658804** [comment]
Hi there.

No, it does not seem to work as expected. After the 1st rightclick, the window is maximized. On the 2nd rightclick, the window is restored but it seems it's being restored at a random position. It's neither the position before the 1st rightclick, nor the position before starting the drag, not even at the center of the screen.

**Comment by RamonUnch** [comment]
Strange, can you give me your ini file?

**Comment by RamonUnch** [comment]
Ok, try with this version of hooks.dll
[hooks1.52test10_x64.zip](https://github.com/RamonUnch/AltSnap/files/8148853/hooks1.52test10_x64.zip)

**Comment by gh658804** [comment]
OK, now it seems to work better, but still...

How it works: press alt+left mouse in a window which is at position A, then rightclick to maximize and again to restore. The window now stays at the position it was before maximizing it (A).

What i expected: press alt+left mouse in a window which is at position A and move it to position B. Without releasing any button, rightclick to maximize then rightclick again to restore. I expected the window to return to position B, now it returns to A.

If the routine could remember position B (which is just before the window was maximized) instead of A (which is just before the window starts moving with Alt+Left-click), it would be perfect for me.

**Comment by RamonUnch** [comment]
> What i expected: press alt+left mouse in a window which is at position A and move it to position B. Without releasing any button, rightclick to maximize then rightclick again to restore. I expected the window to return to position B, now it returns to A.

When using FullWin=1 it should do what you want.
Take Window from A -->move to B, then Rightclick then release everything => the window is maximized.
Again Left+right click => the window is restored to B when FullWin=1 and to A when FullWin=0, this is indeed a bit strange but it is because the window is not moved when using FullWin=1 unless the movement ended. I am not sure this is such a big deal.

**Comment by gh658804** [comment]
FYI, i have FullWin=1 in my INI.

In any case though, what i describe is not exactly the same as what you described. I said "**Without releasing any button**", you said "move to B, then Rightclick **then release everything**", so we are not describing the same situation.

I it's not too much to ask, do you think it's possible to have AltSnap behave as i described in my previous post?

**Comment by RamonUnch** [comment]
Do you mean like the same thing but without the **release everything**? I am not sure why you would want to maximize a window to instantly restore it? anyway the fix could be this:
[hooks1.52text11x64.zip](https://github.com/RamonUnch/AltSnap/files/8149270/hooks1.52text11x64.zip)

**Comment by gh658804** [comment]
Yup. Now it's working as i described! Thanks.

> I am not sure why you would want to maximize a window to instantly restore it?

Well, i guess it's not something one would usually do on purpose, but (for whatever reason) when this happens, IMO he would expect the window to return to its previous position (the one before maximizing it).

One thing i've noticed while testing, is that if you rightclick the window and it gets maximized and then rightclick it again and gets restored, if you still have the Alt + LMB pressed, the window does not move. It gets maximized/restored when you repeat the rightclick but does not move. You have to release everything and press them again for the window to start moving. I'm just wondering if it's the intended behavior?

**Comment by RamonUnch** [comment]
> One thing i've noticed while testing, is that if you rightclick the window and it gets maximized and then rightclick it again and gets restored, if you still have the Alt + LMB pressed, the window does not move. It gets maximized/restored when you repeat the rightclick but does not move. You have to release everything and press them again for the window to start moving. I'm just wondering if it's the intended behavior?

This behavior is necessary in order not to have the window being moved sometime without your cursor even pointing at it:
The latest patch makes it probably happen too much because It makes it even more likely:
Esemple:
1) Move a small window in the bottom right corner of the monitor
2) maximize it (any method), release all keys,
3) Put your cursor in the top left corner of the monitor, then use the Left+right click combo.
The window will be restored in the bottom right corner as intended but if the movement was not blocked then a drag would move the window without having the cursor pointing at it, so this is a no-no for me.

Because the latest patch (test11) tries een more to restore at old position (instead of where the cursor is like older versions), then lock has to happen even more often.
I will rvert the latest test 11 patch because I think it makes it actually worse and a bit too much distant from the original altdrag behavior.

**Comment by RamonUnch** [comment]
I think the best compromise is to make the click combo behaves the new way only when no drag occurred at all.
This way it is mostly compatible with old behaviour and also allows to maximize/restore window at the correct point without having to doubleclick or using the PiercingClick option. Thhis is the case of test 10

**Comment by gh658804** [comment]
> I will rvert the latest test 11 patch because I think it makes it actually worse and a bit too much distant from the original altdrag behavior.

One request on this. Instead of reverting it, could you perhaps make it optional, when a new INI entry is perhaps set to 1?

Well, test 10 did not work that way for me anyway. Also, when a window is maximized and i release everything, it does not gets restored (or moves) when i try to move or restore it...

**Comment by RamonUnch** [comment]
>  Also, when a window is maximized and i release everything, it does not get restored (or moves) when i try to move or restore it... 

Yes, with test 10 the movement gets ignored in both way, but only if no drag started before the right-click.
If you started moving the window, let say to maximize on monitor 2, then you decide you finally want to maximize on monitor 3, thhe you just need to right-click again and move to monitor 3 and again right-click.
With test 11 it gets stuck at each step. I do not wish adding too much options, this click combo feature is already quite obscure.

**Comment by RamonUnch** [comment]
I am not actually sure I understand what you said, It is starting to be a bit complex.

**Comment by gh658804** [comment]
> I do not wish adding too much options, this click combo feature is already quite obscure.

Oh, ok, i understand. Whatever you think is best. The code has to remain maintainable... :)

What i mean is that if i have a maximized window in one monitor and i press Alt + LMB in it,  it does not gets restored and starts moving (as i remember from AltDrag).

**Comment by RamonUnch** [comment]
With test 10:
Alt+LClick+DRAG+RClick => behaves the old way
Alt+LClick+RClick (nodrag) =>toggles the Maximized state but cannot be followed by a drag.
Alt+LClick+DRAG => Restore + moves the the window at the cursor position.

And you are right indeed but this is different with AltSnap:
Optionally you can check the **Restore with single click like original AltDrag** and then:

Alt+LClick (nodrag) =>Restore + moves the the window at the cursor position.

The only reason I added this option is because of AltDrag, initially I was only considering I fixed a bug when windows would be moved without a drag.

**Comment by gh658804** [comment]
> Alt+LClick+DRAG => Restore + moves the the window at the cursor position.

Well, this does not work for me. The window just stays maximized. I tried attaching my INI but it seems there is a problem.

I'm attaching my INI, maybe another setting is interfering...

[AltSnap.zip](https://github.com/RamonUnch/AltSnap/files/8149361/AltSnap.zip)

**Comment by RamonUnch** [comment]
Indeed you checked the **Disable AltSnap on Maximized windows** so in this case you can not AltDrag Maximized window

**Comment by RamonUnch** [comment]
Check the [Advanced Tab]
also the **Restore with single click like original AltDrag** only applies to the snapped window if you checked **Disable AltSnap on Maximized windows **

**Comment by gh658804** [comment]
Ah, great... :( I guess i have to clean my glasses...

Now it's working. Thanks.

**Comment by RamonUnch** [comment]
Thanks for your help, I will keep this issue in my TODO list to see if more improvements can be made in the future
